taskvine/src/worker/vine_worker.c

```diff
@@ -644,19 +680,49 @@ for later processing.
 
 static int handle_completed_tasks(struct link *manager)
 {
-	struct vine_process *p;
+	struct vine_process *p, *pp;
 	uint64_t task_id;
 
 	ITABLE_ITERATE(procs_running, task_id, p)
 	{
-		if (vine_process_is_complete(p)) {
+		/* Skip functions, we take care of them later by checking if the library sends any message. */
+		if (p->type != VINE_PROCESS_TYPE_FUNCTION && vine_process_is_complete(p)) {
 			/* collect the resources associated with the process */
 			reap_process(p, manager);
 
 			/* must reset the table iterator because an item was removed. */
 			itable_firstkey(procs_running);
 		}
+		/* Check if the library has any function invocations done, noblock. */
```

Comment:
This looks like the right idea to me, but please break it out into one or more well named sub-functions.